emulator: Implement SampleRowKeys.

marete opened the pull request:
This implements the SampleRowKeys bigtable RPC.

This change (when combined with the MutateRows PR in an integration branch) fixes the bigtable_table_sample_rows_integration_test  tests which were all failing, and thus fixes TBL-60.

Care has been taken to ensure that the RPC can return any number of sample rows using constant memory.

TESTED=bigtable_table_sample_rows_integration_test now passes

Fixes: TBL-60
References: TBL-60

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8)
<!-- Reviewable:end -->

marete committed:
emulator: Start RowTransaction class.

Logic from Table::MutateRow() will be moved here and this commit
proposes the general structure of this class, for early review.

marete committed:
emulator: Implement undo log and wire-up rewind for the SetCell mutation.

- Introduces a RowTransaction class to proxy mutations and record undo
  logs in case we needs to rewind.
- Re-implements the SetCell mutation in the RowTransaction class.
- Introduces all Undo types/actions for SetCell and records them in an
  undo stack.
- For SetCell only and to facilitate testing with only SetCell
  mutations, introduces the use of RowTransaction in doing the mutation.
- clang-format is run on all affected files (which has removed some
  headers, e.t.c.

A test for SetCell is coming up and is not currently included here.

marete committed:
emulator: Properly declate all possible row mutations.

And unconditionally return an error status for the aggregation
mutations, which we don't currently support.

marete committed:
emulator: Implement a SetCell that records an undo log.

marete committed:
Complete implementation of Undo() for SetCell mutations.

marete committed:
emulator: Implement a constructor for the RowTransaction class.

marete committed:
emulator: Pass RowTransaction a constant reference to the mutation request.

marete committed:
emulator: Introduce the use of a RowTransaction in Table::MutateRow.

Currently in use only for SetCell.

marete committed:
emulator: Run clang-format on the subset of the files we have been changing.

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator.

marete committed:
emulator: Start test suite for row mutation transaction rollback.

marete committed:
emulator: Add a static method to Table that returns a default-constructed table.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: testing: Create a table in the correct way.

Populate a schema protobuf with the minimal fields and call
Table::Create();

marete committed:
emulator: testing: rollback_test: Fix crash when setting up SetCell mutation.

There is a crash later on, in the code that sets up the undo log, that
will be fixed in an upcoming commit.

marete committed:
emulator: SetCell mutation: Various logic fixes for undo log.

marete committed:
emulator: transaction rollback test: Fix crash in destructor.

Use the correct sub-message allocation method, otherwise the protobuf
code becomes confused about who is the owner of the SetCell mutation
message and tries to free it twice, I think.

See also: https://protobuf.dev/reference/cpp/arenas/

marete committed:
emulator: trasnaction rollback test: Fix crash in destructor.

Use the correct sub-message allocation method, otherwise the protobuf
code becomes confused about who is the owner of the SetCell mutation
message and tries to free it twice, I think.

See also: https://protobuf.dev/reference/cpp/arenas/

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator

marete committed:
emulator: table: Add a method to return an iterator to the column families map.

marete committed:
emulator: Introduce methods to retrieve begin and end of column family map iteration.

marete committed:
emulator: Complete implementing test for basic transaction functionality.

The test ensures that SetCell mutations still work via the
RowTransaction class.

Rollback will be tested in an upcoming change.

marete committed:
emulator: rollback test: Fix for more consistent handling of the table shared pointer.

marete committed:
emulator: Store a std::shared_ptr<Table> in RowTransaction.

This is a more correct way of doing it than storing a raw pointer.

marete committed:
emulator: rollback tests: fix test for basic functionality.

Thanks to Marek for the hint on the major part of the fix, which
involved copying a reference instead of a value!

marete committed:
emulator: rollback tests: fix test for basic functionality.

Thanks to Marek for the hint on the major part of the fix, which
involved copying a reference instead of a value!

marete committed:
emulator: ColumnFamily class: Disable copying.

marete committed:
emulator: Disable copying ColumnFamilyRow and simplify tx SetCell logic.

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator

marete committed:
emulator: Disable the copying of ColumnRows.

marete committed:
emulator: rollback_test.cc: factor out helper to check a cell exists with a specific value.

marete committed:
emulator: add helper to set a cell to a value.

Also, run clang-format on the file.

marete committed:
emulator: rollback tests: implement helper function to create tables.

marete committed:
emulator: rollack tests: Fix comment describing test.

marete committed:
emulator: rollback tests: Implement additional testing helper functions.

- To check if a row exists
- To check if a column exists

marete committed:
emulator: add helper to test > 1 SetCell mutations at once.

marete committed:
emulator: use set_cells() to set possibly more than one SetCell mutation.

marete committed:
emualtor: transaction rollback: Implement working test for RestoreValue.

This also involves a fix to the code due to a bug revealed by the
test!

marete committed:
emulator: rollback_tests.cc: run clang-format.

marete committed:
emulator: transaction rollback: Avoid copying the possibly very large cell values.

marete committed:
emulator: transaction rollback: SetCell: Implement working test for DeleteValue.

marete committed:
emulator: transaction rollback: Implement working DeleteRow test.

marete committed:
emulator: transaction rollback: Implement a DeleteFromRow that updates the undo log.

Also, add code coverage output artefacts to ignored files.

marete committed:
emulator: transaction rollback: Implement undo for row deletions (DeleteFromFamily).

And be careful not to potentially very large values (use std::move on
the column qualifier and cell values.

marete committed:
emulator: RestoreRow undo log: Rename struct members more conventionally.

marete committed:
emulator: DeleteFromFamily mutation: Implement with rollback support using RowTransaction.

marete committed:
emulator: transaction rollback: Test basic functionality of a DeleteFromFamily which uses RowTransaction.

marete committed:
emulator: test rollback of a row deletion due to DeleteFromFamily.

And fix a code bug discovered by this test.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: Fixes for review: Improve name of return varible.

In particular, make it clear that we are only examining the status,
and will not use the possibly valid status value (returned object).

marete committed:
emulator: Rename RestoreRow -> RestoreColumnFamilyRow.

marete committed:
emulator: style: Remove trailing underscore from struct member names.

marete committed:
emulator: Rename several iterator variables more appropriately.

marete committed:
emulator: Use std::move() to avoid a copy.

marete committed:
emulator: Rename another iterator properly.

marete committed:
emulator: Use std::move() to avoid another copy.

marete committed:
emulator: check and return from exceptional conditions first.

This is to remain in line with the rest of the code.

marete committed:
emualtor: Fix the naming of another iterator variable.

marete committed:
emulator: remove use of C++17 syntax feature (if with initializer).

We need to have the code compilable by a C++14 compiler.

marete committed:
emulator: rename another iterator variable more appropriately.

marete committed:
emulator: Add a comment justifying storing the column family iterator in the undo log.

In general, it is unsafe to store other kind of map iterators in the
undo log, since a subsequent mutation on the same row could delete the
underlying map, and invalidate the iterator. However, deleting a
column family is a schema operation that cannot happen while we hold
the table lock, therefore for column families only, it is safe to
store the iterators in the undo log.

marete committed:
emulator: transaction rollback: RestoreValue: Eliminate the storage of iterators in undo log.

marete committed:
emulator: actually remove iterator field from RestoreValue.

marete committed:
emulator: Correct and radically simplify SetCell undo code logic.

- Don't store iterators in the DeleteValue undo type (they may be
  invalidated by a subsequent mutation).

- Maintain the invariants that there is never an empty column or empty
  row by using the ColumnFamily::DeleteColumn() function to delete
  ranges of cells during undo. The latter takes care of deleting empty
  columns and empty rows, already.

- And therefore we can get rid of the DeleteRow and DeleteColumn undo
  types and associated code.

All tests continue to pass.

marete committed:
emulator: remove undo types that are no longer needed.

marete committed:
emulator: Delete definitions of rollback types no longer used.

marete committed:
emulator: return unimplemented error status for unsupported aggregation mutations.

Replaces FIXMEs which would silently fail when this mutations are attempted.

marete committed:
emulator: row mutations: use a const reference rather than copying the mutation.

marete committed:
emulator: Modify the Delete* methods to return vector of deleted cells.

And also use std::move() to avoid copying potentially large cell values.

These change is needed to simplify the logic of preparing undo logs
for the Delete* mutations. In particular, we don't have to re-iterate
the table.

marete committed:
emualator: Implement DeleteFromColumn mutation using RowTransaction.

marete committed:
emulator: Test transaction DeleteFromColumn for basic functionality.

marete committed:
emulator: DeleteFromColumn basic functionality test: test that the column is actually deleted.

marete committed:
emulator: make it possible to test a vector of DeleteFromColumn row mutations.

This is needed to test DeleteFromColumn rollback.

marete committed:
emulator: Test transaction rollback for DeleteFromColumn.

marete committed:
emulator: TEST(TransactonRollback, DeleteFromColumnRollback): also check that the cells are properly restored.

marete committed:
emulator: Prepare to support transaction rollbacks for DeleteRow mutation.

In particular, reutrn a map of vectors containing the column qualifers
to columns (cells) deleted, for the undo log.

marete committed:
emulator: DeleteRow: don't forget to erase the row key.

marete committed:
emulator: Fix segmentation fault caused by invalidated iterator in DeleteRow.

Also implement DeleteFromFamily in terms of the new DeleteRow which
returns RestoreValue vectors (and therefore eliminiate the use of one
more undo log type).

marete committed:
emulator: Fix segmentation fault caused by invalidated iterator in DeleteRow.

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator

marete committed:
emulator: clang-format.

marete committed:
emulator: transaction rollback: get rid of another undo log type.

This leaves us with just 2 undo log types: RestoreValue and
DeleteValue, a situation that is much simpler than before.

marete committed:
emulator: Implement a DeleteFromRow that supports rollback.

marete committed:
emulator: Test DeleteFromRow with rollback support and use it in MutateRow code.

marete committed:
emulator: Update and correct a comment.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
erge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
Merge branch 'emulator' of github.com:marete/google-cloud-cpp into emulator

marete committed:
emualtor: Remove now-unused undo type.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: Fixes to avoid copying values that are possibly very large (cell contents).

As well as other fixes for review.

marete committed:
emulator: mutation transactions: use std::chrono:duration_cast.

Makes it easier and safer to change the resolution at a later date.

marete committed:
emulator: use invariant that rows are never empty to eliminate some code.

Preceding code erases all columns. The last column erase in
DeleteColumn will cause the row to be erased. So we don't have to
search for it again in the map and attempt to erase it.

marete committed:
emulator: optimization: make it possible to delete a column via a row iterator.

In this and other situations, this saves us a duplicate call to find
the row by saving and passing an iterator.

marete committed:
emulator: eliminate redundant storage of row_key in undo log.

The entire undo is in the context of a row, therefore all the methods
have access to a row key.

This also reduces the size of the undo log.

marete committed:
emulator: Use google-standard status builder everywhere in main code.

marete committed:
emulator: Add a FIXME to ask to use google-standard Status builder everywhere in test file.

marete committed:
emulator: Remove unused local var.

marete committed:
emulator: eliminate a C++-17ism.

marete committed:
emulator: DeleteRow: fix hang/crash due to use of invalidated row iterator.

marete committed:
emulator: remove non-Google-standard Status builders.

marete committed:
emulator: remove some outdated FIXMEs.

- We now have atomic rollback support for all implemented mutations.

marete committed:
emulator: make some std::chrono usage more type-safe.

marete committed:
emulator: std::move() to avoid another copy of a cell.

marete committed:
emulator: eliminate some code duplication.

marete committed:
emulator: SetCell: return overwritten value (if any).

This prepares the way to simplify the code that prepares the undo log,
and in particular it obviates the need to check to see if the
key (timestamp) existed before the SetCell mutation.

marete committed:
emulator: radically simplify undo log preparation code.

This is thanks to simply returning the old value (if any) rather than
performing almost endless iterations in this function! :)

marete committed:
emulator: ColumnFamily::DeleteRow(): radically simplify the function.

And also eliminate the risk of accessing an invalidated iterator.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: SetCell: set timestamp to server time if passed 0 or negative timestamp.

marete committed:
emulator: transaction rollback: Check that a SetCell with timestamp 0 can be correctly rolled back.

marete committed:
emulator: docs: correct formatting of several comments in header.

marete committed:
emulator: tests: use a better EXPECT* macro in one place.

marete committed:
emulator: Implement ColumnFamily::DeleteTimeStamp().

This allows us to delete cells by specifying a specific timestamp
rather than a timestamp range which might otherwise involve possibly
non-type-safe calculations to determine the upper bound of the
timestamps (in particular doing the latter needs one to assume a
specific resolution for bigtable's timestamp storage).

marete committed:
emulator: transaction undo: radically simplify the code for DeleteValue.

marete committed:
emulator: rollback tests: perform some trivial refactoring.

marete committed:
emulator tests: Fix bigtable_emulator_range_set_test.

The tests aborted due to an assertion in the constructor of Range that
enforced the invariant that start <= end for all well-formed
Ranges. However in several tests testing the ends of ranges, start was
passed in as "unimportant" with end being "A" or "B". This violates
the invariant and causes an abort at runtime.

Fixes: TBL-62.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
emulator: fixes for clang-tidy and clang-format.

Restricted only to changes I have made to this PR.

Tested (all tests pass).

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
Merge branch 'emulator' into fix-test-bigtable_emulator_range_set_test

marete committed:
emulator: SampleRowKeys: Implement a row sampler.

Not yet wired into the GRPC service, and tests against the CI and a
new unit test (not yet implemented) need to be run.

marete committed:
emulator: SampleRowKeys: Wire up the sampler to the SampleRowKeys GRPC call.

This finishes the implementation of the SampleRowKeys RPC.

marete committed:
emulator: SampleRowKeys: Consider the size of the row in all column families.

marete committed:
emulator: SampleRowKeys: Fix crash due to incorrect usage of lambda capture.

Replace incorrect capture of a local variable by reference (which
crashes the emulator due to use of dangling references all over the
place) with a capture by value of a std::shared_ptr which is certain
to be preserved by the capture even when the function surrounding the
lambda returns.

marete committed:
emulator: SampleRowKeys: Fix some typos in an important comment.

marete committed:
emulator: SampleRowKeys: Fix a couple more bugs.

- Fix an infinite loop in returning samples.
- Fix a bug in which the offset of the last row sampled was set to 0.

bigtable_table_sample_rows_integration_test now passes. Yay!

marete committed:
emulator: SampleRowKeys: Revert 2 ci/ files to upstream contents.

These were mistakenly checked in.

marete commented:
I have fixed all the remaining crashes and bugs and now bigtable_table_sample_rows_integration_test tests all pass (you need to merge in the MutateRows PR in an integration branch to verify that).

This PR is now tested and ready for a full review.

PTAL.

marete committed:
emulator: SampleRowKeys: improve the generation of randomness.

- Use more modern means to generate the random
numbers (std::uniform_int_distribution and friends).

- FIX: Seed the random number generator from the random device.

marete committed:
Merge remote-tracking branch 'upstream/emulator' into emulator

marete committed:
Merge branch 'emulator' into sample_row_keys

dopiera reviewed (changes requested):
Quick sanity check before I review the whole code. Can you please tell me why not just create a filtered stream with a row sampler and return the rows? I mean
```
filter.set_row_sample_filter((100/sum_of_the_sizes_of_all_column_families));
auto maybe_stream = CreateFilter(filter, ...);
```
and then just get the row keys of those cells?

The result might be a bit inaccurate but pretty close and the code will be trivial.

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-:-OQ4tIxJ7_0-QUbvS7QM:b-ooxny6)* status: 0 of 4 files reviewed, 2 unresolved discussions

___
*[`google/cloud/bigtable/emulator/table.h` line 156 at r3](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OQ47wMN-tmOY4kshrDN:-OQ47wMO0qxYYfB7935c:b-4o8wj5) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/6898492618352124b4f1dfefb3246f3ba0a239fe/google/cloud/bigtable/emulator/table.h#L156)):*
> ```objc
>           next_sample_closure) {
>     table_ = std::move(table);
>     next_sample_closure_ = std::move(next_sample_closure);
> ```

Please use initialization lists.



_Code quote:_
```objc
    table_ = std::move(table);
    next_sample_closure_ = std::move(next_sample_closure);
```
___
*[`google/cloud/bigtable/emulator/table.h` line 158 at r3](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OQ47_t-92kfT7cilbvW:-OQ47_t-92kfT7cilbvX:bhmg4kl) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/6898492618352124b4f1dfefb3246f3ba0a239fe/google/cloud/bigtable/emulator/table.h#L158)):*
> ```objc
>     next_sample_closure_ = std::move(next_sample_closure);
> 
>     table_->mu_.lock();
> ```

Please use `std::lock_guard` for that (make it a member).


<!-- Sent from Reviewable.io -->

marete committed:
Merge remote-tracking branch 'upstream/emulator' into sample_row_keys

marete committed:
emulator: SampleRowKeys: Start rewrite that uses a filter and CellStream instead.

marete committed:
emulator: trivial mechanical formatting fixes.

marete committed:
emulator: Re-implement SampleRowKeys to use the sampled CellStream.

marete committed:
emulator: server.cc: mechanical formatting fixes.

marete reviewed (commented):
Thanks for pointing that out. My implementation was indeed much more complicated than it should have been.

I have re-written it now to be much simpler by utilizing a bigtable sample row filter to return a sampled CellStream  that can be iterated over, very much like ReadRows.

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-:-ORCCPDF2SQfb0xzIlmG:b6ayxb4)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @dopiera)



<!-- Sent from Reviewable.io -->

marete reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @dopiera)

___
*[`google/cloud/bigtable/emulator/table.h` line 156 at r3](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OQ47wMN-tmOY4kshrDN:-ORCD5oO5UIZmk1G7r_Q:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/6898492618352124b4f1dfefb3246f3ba0a239fe/google/cloud/bigtable/emulator/table.h#L156)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

Please use initialization lists.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.h` line 158 at r3](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OQ47_t-92kfT7cilbvW:-ORCD6Zc4tB3oRppz8_l:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/6898492618352124b4f1dfefb3246f3ba0a239fe/google/cloud/bigtable/emulator/table.h#L158)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

Please use `std::lock_guard` for that (make it a member).
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

marete reviewed (commented):
I have re-written it now to be much simpler by utilizing a bigtable sample row filter to return a sampled CellStream that can be iterated over, very much like ReadRows.

PTAL.

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-:-ORCDEbx2-XGDKEK0oKm:b-4w8ajo)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @dopiera)



<!-- Sent from Reviewable.io -->

marete commented:
PTAL.

dopiera reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8)* status: 0 of 4 files reviewed, 4 unresolved discussions (waiting on @marete)

___
*[`google/cloud/bigtable/emulator/server.cc` line 58 at r6](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-ORDPrKh1x7Adrmb6Rlt:-ORDPrKh1x7Adrmb6Rlu:bgvanfg) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/ef4d3c2b7f091903893d1a4716da7473264ef5b9/google/cloud/bigtable/emulator/server.cc#L58)):*
> ```cpp
>   }
> 
>   grpc::Status SampleRowKeys(
> ```

I like the idea in general, but I think I'd rework this function. I.e. I'd do the following:
* estimate row size based on the first row, i.e.
  * create an unfiltered stream
  * iterate through all cells
    * accumulate the sizes of all rows, CF, column qualifiers, timestamps and values
  * memoize the estimate
  * create the filtered stream
    * proceed as you do, but use the memoized row size for estimates



_Code quote:_
```cpp
grpc::Status SampleRowKeys(
```
___
*[`google/cloud/bigtable/emulator/server.cc` line 71 at r6](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-ORDNNDF6IHK9je4T28n:-ORDNNDF6IHK9je4T28o:b-ollbw9) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/ef4d3c2b7f091903893d1a4716da7473264ef5b9/google/cloud/bigtable/emulator/server.cc#L71)):*
> ```cpp
>     // Return ~ 1/100 rows.
>     auto maybe_stream =
>         table->GetSampledRowsCellStream(static_cast<double>(1) / 100);
> ```

I think I'd add at least two more orders of magnitude here. BTW, what wrong with `0.0001` notation? It will automatically get cast to `double`



_Code quote:_
```cpp
static_cast<double>(1) / 100
```
___
*[`google/cloud/bigtable/emulator/server.cc` line 89 at r6](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-ORDM_pR0gzCCDgX0vsB:-ORDM_pS7bbRkwnszYBT:b-vv8dma) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/ef4d3c2b7f091903893d1a4716da7473264ef5b9/google/cloud/bigtable/emulator/server.cc#L89)):*
> ```cpp
> 
>       // This is an estimate
>       offset_bytes += (100 * row_size);
> ```

The estimate is off I'm afraid. Here, you're taking the size of a single cell, but there may be many in this row.

What you probably want to do is to iterate through all cells of this row instead.



_Code quote:_
```cpp
      offset_bytes += (100 * row_size);
```
___
*[`google/cloud/bigtable/emulator/server.cc` line 99 at r6](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-ORDPddu7NCs-VxLQ-LD:-ORDPddv8x_RGLmdRzYA:b-61ignm) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/ef4d3c2b7f091903893d1a4716da7473264ef5b9/google/cloud/bigtable/emulator/server.cc#L99)):*
> ```cpp
> 
>     // We need to return at least one row if the table is not empty;
>     if (!offset_bytes) {
> ```

Why can't we return the empty string?



_Code quote:_
```cpp
    if (!offset_bytes) {
```


<!-- Sent from Reviewable.io -->

marete committed:
Merge branch 'upstream_emulator' into sample_row_keys

marete committed:
emulator: SampleRowKeys: Re-write to compute exact offset for row keys.

marete reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8)* status: 0 of 4 files reviewed, 4 unresolved discussions (waiting on @dopiera and @marete)

___
*[`google/cloud/bigtable/emulator/server.cc` line 58 at r6](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-ORDPrKh1x7Adrmb6Rlt:-ORSm2_n4hqe-CDKNjO2:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/ef4d3c2b7f091903893d1a4716da7473264ef5b9/google/cloud/bigtable/emulator/server.cc#L58)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

I like the idea in general, but I think I'd rework this function. I.e. I'd do the following:
* estimate row size based on the first row, i.e.
  * create an unfiltered stream
  * iterate through all cells
    * accumulate the sizes of all rows, CF, column qualifiers, timestamps and values
  * memoize the estimate
  * create the filtered stream
    * proceed as you do, but use the memoized row size for estimates
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/server.cc` line 71 at r6](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-ORDNNDF6IHK9je4T28n:-ORSm4Dx-EqeRE6k3jJG:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/ef4d3c2b7f091903893d1a4716da7473264ef5b9/google/cloud/bigtable/emulator/server.cc#L71)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

I think I'd add at least two more orders of magnitude here. BTW, what wrong with `0.0001` notation? It will automatically get cast to `double`
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/server.cc` line 89 at r6](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-ORDM_pR0gzCCDgX0vsB:-ORSm5ZK4vOECO76EtGh:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/ef4d3c2b7f091903893d1a4716da7473264ef5b9/google/cloud/bigtable/emulator/server.cc#L89)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

The estimate is off I'm afraid. Here, you're taking the size of a single cell, but there may be many in this row.

What you probably want to do is to iterate through all cells of this row instead.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/server.cc` line 99 at r6](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-ORDPddu7NCs-VxLQ-LD:-ORSm6va-A8oxdLGWjNH:bg9nbsl) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/ef4d3c2b7f091903893d1a4716da7473264ef5b9/google/cloud/bigtable/emulator/server.cc#L99)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

Why can't we return the empty string?
</blockquote></details>

Because the client testing code expects at least one sample if it populated the table with any rows at all. See the comment in the code and also the test.


<!-- Sent from Reviewable.io -->

marete reviewed (commented):
All comments implemented. PTAL. Thanks!

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-:-ORSmvzz9nDVL4rjvQTK:b-f96v1x)* status: 0 of 4 files reviewed, 4 unresolved discussions (waiting on @dopiera and @marete)



<!-- Sent from Reviewable.io -->

marete committed:
emulator: Remove unused code from earlier PR iteration.

marete committed:
emulator: mechanical fixes of typos.

marete committed:
emulator: clang-tidy fixes.

marete committed:
Merge branch 'upstream_emulator' into sample_row_keys

marete committed:
Merge branch 'upstream_emulator' into sample_row_keys

marete committed:
Merge branch 'upstream_emulator' into sample_row_keys

marete committed:
emulator: mechanical fixes by CI tool.

marete committed:
Merge branch 'upstream_emulator' into sample_row_keys

dopiera reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8)* status: 0 of 5 files reviewed, 3 unresolved discussions (waiting on @marete)

___
*[`google/cloud/bigtable/emulator/server.cc` line 58 at r6](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-ORDPrKh1x7Adrmb6Rlt:-OTkPESAEHWgYu5rRdwV:b-fg1hnt) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/ef4d3c2b7f091903893d1a4716da7473264ef5b9/google/cloud/bigtable/emulator/server.cc#L58)):*
I don't get it. What you seem doing is accumulate the size of all column family names and all column qualifier names within a row only to sum them afterwards.

So if the row looks like this:
```
"row1", "cf1:col1", "some_value"
"row1", "cf1:col2", "some_value"
"row1", "cf2:col1", "some_value"
```
You'll add to row offset `len(cf1) + len(cf2) + len(col1) + len(col2) = 14`

This, I think, will not really reflect the amount of space this row takes. We need to make a decision on what we really want to count and I see three choices:
1. try to estimate how big our in-memory structures are
2. for every CellView returned by the stream, sum the sizes of all its
  fields (row_key, column_family, column_qualifier, value and
  sizeof(timestamp)
3. try to estimate how big SSTables would have been had we used them to
  store the data

I'm guessing that your intention was to estimate option 1. I think the example I've shown exemplifies the bug - you only count `col1` once but you should count it twice - once for `cf1` and once for `cf2`. Also, you should not count the column families at all - they are not repeated for every row - there is one column family name string per whole table.

The comment in the proto states:
> Approximate total storage space used by all rows in the table which precede `row_key`. Buffering the contents of all rows between two subsequent samples would require space roughly equal to the difference in their `offset_bytes` fields.

I think we can infer from it that the dumbest option (number 2) is the one we should go with.

You're also traversing the whole table (that's the only way to compute the size). My suggestion in the comment was to only estimate the size of the first row and use that for calculations instead.
___
*[`google/cloud/bigtable/emulator/table.cc` line 605 at r10](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OTkPnjwCfpVC2d0_oHy:-OTkPnjx7_MNOILrij8b:by5w8qz) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/94cdc27bdee9558eabf0bcf74df5e6c8c53017bc/google/cloud/bigtable/emulator/table.cc#L605)):*
> ```cpp
> 
>   for (; stream; ++stream) {
>     auto row_key = stream->row_key();
> ```

Unnecessary copy, I think.
___
*[`google/cloud/bigtable/emulator/table.cc` line 626 at r10](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OTkPrTAATD5cIl4u-Lq:-OTkPrTAATD5cIl4u-Lr:b-doz4x) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/94cdc27bdee9558eabf0bcf74df5e6c8c53017bc/google/cloud/bigtable/emulator/table.cc#L626)):*
> ```cpp
>       current_row_key = row_key;
> 
>       first_row = false;
> ```

This if first_cell, really, no? Also, I don't think this is needed at all.


<!-- Sent from Reviewable.io -->

marete committed:
Merge branch 'upstream_emulator' into sample_row_keys

marete committed:
Merge branch 'upstream_emulator' into sample_row_keys

marete committed:
Merge branch 'upstream_emulator' into sample_row_keys

marete committed:
perf: emulator: SampleRowKeys: Radically simplify the code.

- There is no need to read the entire table

Actually, after reading a couple of rows to estimate size, there is no
need to read any table data at all (though that is an optimization we
will make later).

marete committed:
style: emulator: mechanical fixes by CI tool.

marete committed:
fix: emulator: Fix the use of an uninitialized variable.

marete commented:
PTAL. Thanks.

marete committed:
fix: emulator: SampleRowKeys: Eliminate a needless copy.

marete reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8)* status: 0 of 5 files reviewed, 3 unresolved discussions (waiting on @dopiera and @prawilny)

___
*[`google/cloud/bigtable/emulator/server.cc` line 58 at r6](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-ORDPrKh1x7Adrmb6Rlt:-OUM3DQG5Mkh3rXZMOAj:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/ef4d3c2b7f091903893d1a4716da7473264ef5b9/google/cloud/bigtable/emulator/server.cc#L58)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

I don't get it. What you seem doing is accumulate the size of all column family names and all column qualifier names within a row only to sum them afterwards.

So if the row looks like this:
```
"row1", "cf1:col1", "some_value"
"row1", "cf1:col2", "some_value"
"row1", "cf2:col1", "some_value"
```
You'll add to row offset `len(cf1) + len(cf2) + len(col1) + len(col2) = 14`

This, I think, will not really reflect the amount of space this row takes. We need to make a decision on what we really want to count and I see three choices:
1. try to estimate how big our in-memory structures are
2. for every CellView returned by the stream, sum the sizes of all its
  fields (row_key, column_family, column_qualifier, value and
  sizeof(timestamp)
3. try to estimate how big SSTables would have been had we used them to
  store the data

I'm guessing that your intention was to estimate option 1. I think the example I've shown exemplifies the bug - you only count `col1` once but you should count it twice - once for `cf1` and once for `cf2`. Also, you should not count the column families at all - they are not repeated for every row - there is one column family name string per whole table.

The comment in the proto states:
> Approximate total storage space used by all rows in the table which precede `row_key`. Buffering the contents of all rows between two subsequent samples would require space roughly equal to the difference in their `offset_bytes` fields.

I think we can infer from it that the dumbest option (number 2) is the one we should go with.

You're also traversing the whole table (that's the only way to compute the size). My suggestion in the comment was to only estimate the size of the first row and use that for calculations instead.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.cc` line 605 at r10](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OTkPnjwCfpVC2d0_oHy:-OUM9TiO2triziQXBZV0:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/94cdc27bdee9558eabf0bcf74df5e6c8c53017bc/google/cloud/bigtable/emulator/table.cc#L605)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

Unnecessary copy, I think.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.cc` line 626 at r10](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OTkPrTAATD5cIl4u-Lq:-OUM9S6K85TdPJtzi0Sd:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/94cdc27bdee9558eabf0bcf74df5e6c8c53017bc/google/cloud/bigtable/emulator/table.cc#L626)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

This if first_cell, really, no? Also, I don't think this is needed at all.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

marete reviewed (commented):
Thank you for your careful review. PTAL.

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-:-OUM9ZI5AlmiWtL6NMPi:b-hgpwmx)* status: 0 of 5 files reviewed, 3 unresolved discussions (waiting on @dopiera and @prawilny)



<!-- Sent from Reviewable.io -->

prawilny reviewed (commented):
In general it looks good, I only left some minor comments.

> Reviewed 1 of 3 files at r9, 3 of 4 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-:-OUZvNZH7FSJ1BKVSqbK:b-8pywvx)* status: 4 of 5 files reviewed, 14 unresolved discussions (waiting on @dopiera)

___
*[`google/cloud/bigtable/emulator/table.cc` line 28 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZuZoXEU2Ho-pWC5mx:-OUZuZoXEU2Ho-pWC5my:buhg5aq) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L28)):*
> ```cpp
> #include <google/bigtable/v2/bigtable.pb.h>
> #include <google/bigtable/v2/data.pb.h>
> #include <absl/strings/ascii.h>
> ```

[nit] My IDE claims that we don't use it
___
*[`google/cloud/bigtable/emulator/table.cc` line 35 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZufor2MWPaL65_dwE:-OUZufor2MWPaL65_dwF:buhg5aq) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L35)):*
> ```cpp
> #include <re2/re2.h>
> #include <chrono>
> #include <climits>
> ```

[nit] My IDE claims that we don't use it
___
*[`google/cloud/bigtable/emulator/table.cc` line 37 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZuh_g26x8vMj47dqY:-OUZuh_g26x8vMj47dqZ:buhg5aq) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L37)):*
> ```cpp
> #include <climits>
> #include <cmath>
> #include <cstddef>
> ```

[nit] My IDE claims that we don't use it
___
*[`google/cloud/bigtable/emulator/table.cc` line 45 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZul22-thFSqVzAPFC:-OUZul22-thFSqVzAPFD:buhg5aq) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L45)):*
> ```cpp
> #include <optional>
> #include <random>
> #include <ratio>
> ```

[nit] My IDE claims that we don't use it
___
*[`google/cloud/bigtable/emulator/table.cc` line 637 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZvDlK93Lcm2xUpWMw:-OUZvDlK93Lcm2xUpWMx:b-xdeosa) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L637)):*
> ```cpp
>   auto& stream = *maybe_all_rows_steam;
> 
>   absl::optional<std::string> current_row_key;
> ```

[nit] maybe `first_row_key` instead?
___
*[`google/cloud/bigtable/emulator/table.cc` line 654 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZjpQD8_SwsJC9ojeh:-OUZjpQD8_SwsJC9ojei:b-gf9h4q) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L654)):*
> ```cpp
>     current_row_key = stream->row_key();
> 
>     row_size_estimate += stream->row_key().size();
> ```

Shouldn't we count the row key only once?
___
*[`google/cloud/bigtable/emulator/table.cc` line 689 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZtx8z2dnHZm5tPuc_:-OUZtx8z2dnHZm5tPuca:b-lxa37i) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L689)):*
> ```cpp
>   bool wrote_a_sample;
> 
>   for (; sampled_stream; ++sampled_stream) {
> ```

I think that if you are unlucky, you might get a duplicate row key like this, because the stream streams cells rather than rows.
If you use `sampled_stream.Next(kRow)`, you are guaranteed to avoid this problem.
___
*[`google/cloud/bigtable/emulator/table.cc` line 701 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZoVCL7JEc7o5LP1sW:-OUZoVCL7JEc7o5LP1sX:b76h1ap) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L701)):*
> ```cpp
> 
>   // Cloud bigtable client tests expect that, if they populated the
>   // table with at least one row, then at least one row sampele is
> ```

sampele => sample
___
*[`google/cloud/bigtable/emulator/table.cc` line 707 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZpM9H2_vm8TuY6Hga:-OUZpM9H2_vm8TuY6Hgb:b-xlmozg) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L707)):*
> ```cpp
>   // and an offset that is the estimated row size * the number of rows
>   // in the largest column family. We can return any string without
>   // lookup because the keys returned need not ever have been written
> ```

"the keys returned need not ever have been written to" - I am unsure of my understanding of this sentence. Please reword it.

Maybe just use something like "the returned keys don't have to exist in the table"?
___
*[`google/cloud/bigtable/emulator/table.cc` line 710 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZt6be37xI-4wbEmPA:-OUZt6be37xI-4wbEmPB:b-uu26ql) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L710)):*
> ```cpp
>   // to.
>   if (!wrote_a_sample) {
>     std::size_t max = 0;
> ```

[nit] Please rename it to something like `row_number_estimate` - it took me a while to why it all is okay and I think such a renaming would help the next reader with that.
___
*[`google/cloud/bigtable/emulator/table.cc` line 723 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZpvx3-9_yvt7BgUyW:-OUZpvx3-9_yvt7BgUyX:b-6hvn3o) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L723)):*
> ```cpp
>     writer->Write(std::move(resp));
> 
>     google::bigtable::v2::SampleRowKeysResponse last_resp;
> ```

I think we can replace this whole block from this line on with `offset = max * row_size_estimate;` (if we do, it might be better to have an intermediary variable and use it both in set_offset_bytes() and `offset = ...` assignment).
It would make it clearer that the special case is only writing a single response stream entry, not also closing the stream.


<!-- Sent from Reviewable.io -->

marete committed:
Merge branch 'upstream_emulator' into sample_row_keys

marete committed:
style: emulator: headers: clang-tidy cleanup of unused and missing headers.

marete committed:
style: emulator: Fix typo in identifier.

marete committed:
fix: SampleRowKeys: count the row_key just once and name a variable more appropriately.

marete committed:
fix: SampleRowKeys: ensure we don't sample more than once.

Advance the stream by rows so we don't risk streaming the same row key
more than once.

Also other changes for review.

marete committed:
style: SampleRowKeys: further code simplification.

marete committed:
style: mechanical formatting fixes.

marete reviewed (commented):
PTAL.

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-:-OUaCbW99Lx_WsOdABEw:b-83330m)* status: 2 of 5 files reviewed, 14 unresolved discussions (waiting on @dopiera and @prawilny)

___
*[`google/cloud/bigtable/emulator/table.cc` line 28 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZuZoXEU2Ho-pWC5mx:-OU_odwD-9yduFjb9kRY:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L28)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

[nit] My IDE claims that we don't use it
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.cc` line 35 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZufor2MWPaL65_dwE:-OU_oeh51Xy0Ay26MUD4:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L35)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

[nit] My IDE claims that we don't use it
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.cc` line 37 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZuh_g26x8vMj47dqY:-OU_ogB7C9WmMd-lESZz:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L37)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

[nit] My IDE claims that we don't use it
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.cc` line 45 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZul22-thFSqVzAPFC:-OU_ogeM9PhhEH5HZDTF:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L45)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

[nit] My IDE claims that we don't use it
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.cc` line 637 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZvDlK93Lcm2xUpWMw:-OU_rN0L1cp8K_VUfOLr:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L637)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

[nit] maybe `first_row_key` instead?
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.cc` line 654 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZjpQD8_SwsJC9ojeh:-OU_xYqgAwyoqW3YyW-B:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L654)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

Shouldn't we count the row key only once?
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.cc` line 689 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZtx8z2dnHZm5tPuc_:-OUa1iBFD0RgaGC1azIM:bnjn4iz) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L689)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

I think that if you are unlucky, you might get a duplicate row key like this, because the stream streams cells rather than rows.
If you use `sampled_stream.Next(kRow)`, you are guaranteed to avoid this problem.
</blockquote></details>

Thank you for this excellent insight. Fixed.
___
*[`google/cloud/bigtable/emulator/table.cc` line 701 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZoVCL7JEc7o5LP1sW:-OU_xflJ7zn_5awkbUQn:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L701)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

sampele => sample
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.cc` line 707 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZpM9H2_vm8TuY6Hga:-OU_yKYj18M3vNI6NNza:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L707)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

"the keys returned need not ever have been written to" - I am unsure of my understanding of this sentence. Please reword it.

Maybe just use something like "the returned keys don't have to exist in the table"?
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.cc` line 710 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZt6be37xI-4wbEmPA:-OU_zDq7BUORX5JJL9Xe:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L710)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

[nit] Please rename it to something like `row_number_estimate` - it took me a while to why it all is okay and I think such a renaming would help the next reader with that.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.cc` line 723 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZpvx3-9_yvt7BgUyW:-OUaBAV0-BuCx77ZGGA8:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L723)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

I think we can replace this whole block from this line on with `offset = max * row_size_estimate;` (if we do, it might be better to have an intermediary variable and use it both in set_offset_bytes() and `offset = ...` assignment).
It would make it clearer that the special case is only writing a single response stream entry, not also closing the stream.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

prawilny reviewed (approved):
Please fix this one comment about estimating row size before merging.

> Reviewed 2 of 3 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-:-OUcZN4w4m3mpNyFkKlH:b-2iu6oo)* status: 4 of 5 files reviewed, 5 unresolved discussions (waiting on @dopiera and @marete)

___
*[`google/cloud/bigtable/emulator/table.cc` line 35 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZufor2MWPaL65_dwE:-OUcYeyK387PEG1EXPmb:b-f34e1l) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L35)):*
<details><summary><i>Previously, marete (Brian Gitonga Marete) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

FYI my IDE still complains about it.
___
*[`google/cloud/bigtable/emulator/table.cc` line 654 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZjpQD8_SwsJC9ojeh:-OUcWjgtAMbrofDIKk1J:b-l0twus) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L654)):*
<details><summary><i>Previously, marete (Brian Gitonga Marete) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Sorry, when writing this comment I thought that we do store the row key only once, but in the meantime I realized it was wrong - we use it as a key in each column, so unfortunately I think we should revert this change and count it as before - for every cell.

Sorry for that!
___
*[`google/cloud/bigtable/emulator/table.cc` line 44 at r13](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUcYn58-cID0lwkFM_5:-OUcYn58-cID0lwkFM_6:b-ibyy9p) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/391dcbead026b2c393d68d9711c8e706c48a64b1/google/cloud/bigtable/emulator/table.cc#L44)):*
> ```cpp
> #include <cstdint>
> #include <cstdlib>
> #include <functional>
> ```

FYI my IDE complains about it. It's already included in the `table.h` header.


<!-- Sent from Reviewable.io -->

marete committed:
fix: SampleRowKeys: count the row_key multiple times.

marete reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8)* status: 4 of 5 files reviewed, 5 unresolved discussions (waiting on @dopiera and @prawilny)

___
*[`google/cloud/bigtable/emulator/table.cc` line 35 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZufor2MWPaL65_dwE:-OUe87VcDKNDGLwB_diu:b-ofci0y) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L35)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

FYI my IDE still complains about it.
</blockquote></details>

My IDE has this feature and does not complain about it. I want to leave it to a PR of its own, mechanically generated by clang-tidy.
___
*[`google/cloud/bigtable/emulator/table.cc` line 654 at r12](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/8#-OUZjpQD8_SwsJC9ojeh:-OUe8K6_3deMGULLli2E:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/1020f9566986c89e1b0d98050049ac6f3277a5fa/google/cloud/bigtable/emulator/table.cc#L654)):*
<details><summary><i>Previously, prawilny (Adam Czajkowski) wrote…</i></summary><blockquote>

Sorry, when writing this comment I thought that we do store the row key only once, but in the meantime I realized it was wrong - we use it as a key in each column, so unfortunately I think we should revert this change and count it as before - for every cell.

Sorry for that!
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
